packages/notebook/src/history.ts

```diff
@@ -0,0 +1,419 @@
+      for (let i = 0; i < value.content.history.length; i++) {
+        current = value.content.history[i] as string[];
+        if (current !== last) {
+          kernelSession = (value.content.history[i] as string[])[0];
+          this._history.push((last = current));
+        }
+      }
+      // set the kernel session for filtering
+      if (!this.kernelSession) {
+        if (current[2] == cell?.model.sharedModel.getSource()) {
+          this._kernelSession = kernelSession;
+        }
+      }
+    }
+  }
+
+  /**
+   * Handle a text change signal from the editor.
+   */
+  protected onTextChange(): void {
+    if (this._setByHistory) {
+      this._setByHistory = false;
+      return;
+    }
+    this.reset();
+  }
+
+  /**
+   * Handle the current kernel changing.
+   */
+  private async _handleKernel(): Promise<void> {
+    this._kernel = this.sessionContext.session?.kernel;
+    if (!this._kernel) {
+      this._history.length = 0;
+      return;
+    }
+    await this._retrieveHistory().catch();
+    return;
+  }
+
+  /**
+   * retrieve the history from the kernel
+   *
+   * @param cell - The string to use when filtering the data.
+   */
+  private async _retrieveHistory(cell?: Cell): Promise<void> {
+    return await this._kernel
+      ?.requestHistory(request(this.toRequest))
+      .then(v => {
+        this.onHistory(v, cell);
+      })
+      .catch(() => {
+        console.warn(
+          this.translator
+            .load('jupyterlab')
+            .__('History was unable to be retrieved')
+        );
+      });
+  }
+
+  /**
+   * Set the filter data.
+   *
+   * @param filterStr - The string to use when filtering the data.
+   */
+  protected setFilter(filterStr: string = ''): void {
+    // Apply the new filter and remove contiguous duplicates.
+    this._filtered.length = 0;
+
+    let last = '';
+    let current = '';
+    for (let i = 0; i < this._history.length; i++) {
+      current = this._history[i][2] as string;
+      if (current !== last && filterStr !== current) {
+        this._filtered.push((last = current));
+      }
+    }
+    this._filtered.push(filterStr);
+  }
+
+  private _cursor = 0;
+  private _hasSession = false;
+  private _history: Array<Array<string>> = [];
+  private _placeholder: string = '';
+  private _kernelSession: string = '';
+  private _setByHistory = false;
+  private _isDisposed = false;
+  private _editor: CodeEditor.IEditor | null = null;
+  private _filtered: string[] = [];
+  private _kernel: IKernelConnection | null | undefined = null;
+}
+
+/**
+ * A namespace for NotebookHistory statics.
+ */
+export namespace NotebookHistory {
+  /**
+   * The initialization options for a console history object.
+   */
+  export interface IOptions {
+    /**
+     * The client session used by the foreign handler.
+     */
+    sessionContext: ISessionContext;
+
+    /**
+     * The application language translator.
+     */
+    translator: ITranslator;
```

Comment:
The constructor supports optional translator but not this interface:

```suggestion
    translator?: ITranslator;
```